mypy/checker.py

```diff
@@ -3483,21 +3484,34 @@ def find_isinstance_check(self, node: Expression
                     vartype = type_map[expr]
                     return self.conditional_callable_type_map(expr, vartype)
         elif isinstance(node, ComparisonExpr):
-            # Check for `x is None` and `x is not None`.
+            operand_types = [coerce_to_literal(type_map[expr])
+                             for expr in node.operands if expr in type_map]
+
             is_not = node.operators == ['is not']
-            if any(is_literal_none(n) for n in node.operands) and (
-                    is_not or node.operators == ['is']):
+            if (is_not or node.operators == ['is']) and len(operand_types) == len(node.operands):
                 if_vars = {}  # type: TypeMap
                 else_vars = {}  # type: TypeMap
-                for expr in node.operands:
-                    if (literal(expr) == LITERAL_TYPE and not is_literal_none(expr)
-                            and expr in type_map):
+
+                for i, expr in enumerate(node.operands):
+                    var_type = operand_types[i]
+                    other_type = operand_types[1 - i]
```

Comment:
We've actually never properly handled this case, I think. The old if-check on line 3490-ish lets this code run only if there's exactly only a single operator; the new if-check I'm replacing that with continues to do the same thing. So, as a consequence, we can safely assume there'll be exactly two operands at this point.

I have a fix for this, but I decided it might be better to submit it as a separate PR. Once I combined this with the equality changes mentioned above, the changes ended up being much more intrusive.

Reply:
> The old if-check on line 3490-ish lets this code run only if there's exactly only a single operator

Maybe I am missing something, but the code there looks like it is about completely different case, it is about `isinstance()` and `issubclass()` having other number of arguments than two.

(Also it is in a different `if` branch, so it will not affect this branch).

Reply:
Oh, maybe the line numbers shifted after I merged. It's now line 3541-ish.

The old check used to do this:
    
    is_not = node.operators == ['is not']
    if any(is_literal_none(n) for n in node.operands) and (
                    is_not or node.operators == ['is']):

And the new checks do this:

    is_not = node.operators == ['is not']
    if (is_not or node.operators == ['is']) and len(operand_types) == len(node.operands):

We also make the same assumption when handling the `==` and `in` operators as well -- those are:

    elif node.operators == ['==']:

and:

    elif node.operators in [['in'], ['not in']]: